src/MySqlConnector/Core/Context.cs

```diff
@@ -0,0 +1,29 @@
+using MySqlConnector.Protocol;
+
+namespace MySqlConnector.Core;
+
+internal sealed class Context
+{
+	public Context(ProtocolCapabilities protocolCapabilities, string? database, int connectionId)
+	{
+		SupportsDeprecateEof = (protocolCapabilities & ProtocolCapabilities.DeprecateEof) != 0;
+		SupportsCachedPreparedMetadata = (protocolCapabilities & ProtocolCapabilities.MariaDbCacheMetadata) != 0;
+		SupportsQueryAttributes = (protocolCapabilities & ProtocolCapabilities.QueryAttributes) != 0;
+		SupportsSessionTrack = (protocolCapabilities & ProtocolCapabilities.SessionTrack) != 0;
+		ConnectionId = connectionId;
+		Database = database;
+		m_initialDatabase = database;
+	}
+
+	public bool SupportsDeprecateEof { get; }
+	public bool SupportsQueryAttributes { get; }
+	public bool SupportsSessionTrack { get; }
+	public bool SupportsCachedPreparedMetadata { get; }
+	public string? ClientCharset { get; set; }
+
+	public string? Database { get; set; }
```

Comment:
I see the appeal in moving the connection-context-specific data into this new class, but I think it makes it harder to understand to have state spread across `ServerSession` and now `Context`. I'd rather this class be an immutable collection of data that's passed into methods that need to query the current capabilities.

I also think it's a little "unpredictable" to have these properties updated without any notification that it's happening. (The old way wasn't great, either, in that code had to read the `OkPayload.NewSchema` and act on it, and if it didn't, that was a bug.) I also don't like the OkPayload parsing code having the side-effect of updating this mutable state, instead of just being a property-bag/DTO of all the data that was read out of the OK packet.

Made it immutable here: https://github.com/mysql-net/MySqlConnector/pull/1497/commits/77ad1910007923da2a6f12244eaf01eb30c77dcd